cpp/ql/src/semmle/code/cpp/Declaration.qll

```diff
@@ -193,20 +193,87 @@ abstract class Declaration extends Locatable, @declaration {
 
   /**
    * Gets a template argument used to instantiate this declaration from a template.
-   * When called on a template, this will return a template parameter.
+   * When called on a template, this will return a template parameter type for
+   * both typed and non-typed parameters.
    */
-  final Type getATemplateArgument() { result = getTemplateArgument(_) }
+  final Locatable getATemplateArgument() { result = getTemplateArgument(_) }
+
+  /**
+   * Gets a template argument used to instantiate this declaration from a template.
+   * When called on a template, this will return a non-typed template
+   * parameter value.
+   */
+  final Locatable getATemplateArgumentKind() { result = getTemplateArgumentKind(_) }
 
   /**
    * Gets the `i`th template argument used to instantiate this declaration from a
-   * template. When called on a template, this will return the `i`th template parameter.
+   * template.
+   *
+   * For example:
+   *
+   * `template<typename T, T X> class Foo;`
+   *
+   * Will have `getTemplateArgument(0)` return `T`, and
+   * `getTemplateArgument(1)` return `X`.
+   *
+   * `Foo<int, 1> bar;
+   *
+   * Will have `getTemplateArgument())` return `int`, and
+   * `getTemplateArgument(1)` return `1`.
    */
-  Type getTemplateArgument(int index) { none() }
+  final Locatable getTemplateArgument(int index) {
+    if exists(getTemplateArgumentValue(index))
+    then result = getTemplateArgumentValue(index)
+    else result = getTemplateArgumentType(index)
+  }
+
+  /**
+   * Gets the `i`th template argument value used to instantiate this declaration
+   * from a template. When called on a template, this will return the `i`th template
+   * parameter value if it exists.
+   *
+   * For example:
+   *
+   * `template<typename T, T X> class Foo;`
+   *
+   * Will have `getTemplateArgumentKind(1)` return `T`, and no result for
+   * `getTemplateArgumentKind(0)`.
+   *
+   * `Foo<int, 10> bar;
+   *
+   * Will have `getTemplateArgumentKind(1)` return `int`, and no result for
+   * `getTemplateArgumentKind(0)`.
+   */
+  final Locatable getTemplateArgumentKind(int index) {
+    if exists(getTemplateArgumentValue(index))
+    then result = getTemplateArgumentType(index)
+    else none()
+  }
 
   /** Gets the number of template arguments for this declaration. */
   final int getNumberOfTemplateArguments() {
     result = count(int i | exists(getTemplateArgument(i)))
   }
+
+  /**
+   * INTERNAL: Do not use.
+   *
+   * Gets a Type for a template argument.  May be the template argument itself
+   * or the type of a non-type template argument.
+   *
+   * Use `getTemplateArgument` or `getTemplateKind` instead.
+   */
+  Type getTemplateArgumentType(int index) { none() }
+
+  /**
+   * INTERNAL: Do not use.
+   *
+   * Gets an Expression representing the value of a non-type template
+   * argument.
+   *
+   * Use `getTemplateArgument` or `getTemplateKind` instead.
+   */
+  Expr getTemplateArgumentValue(int index) { none() }
```

Comment:
I don't see the value in using QL dynamic dispatch for `getTemplateArgumentValue` and `getTemplateArgumentType`. It confuses the auto-complete list and the auto-generated documentation. If you make `getTemplateArgumentValue` `private`, couldn't its body just be:
```
     class_template_argument_value(underlyingElement(this), index, unresolveElement(result))
     or
     function_template_argument_value(underlyingElement(this), index, unresolveElement(result))
     or
     variable_template_argument_value(underlyingElement(this), index, unresolveElement(result))
```
and similar for `getTemplateArgumentType`?

Reply:
I'm happy to change this - I thought it looked odd when I did the initial changes, but followed the preexisting code pattern.

The question then is - why do we have three underlying tables?  Would it make more sense to merge them all into one table?  Would that improve performance.

Reply:
I have no opinion on that. But I don't think performance will ever be a concern for these tables.

Reply:
I've made the changes you suggested.

Reply:
Looks good. Except I think the qlformat test will fail because of the indentation of `or`.